### ISSUE TITLE
Test failures for 2020.0.0-beta1: b_jobs/b_times

### ISSUE THREAD

**Issue by eli-schwartz** [att/ast#1398]
**Description of problem:**

```
147/286 b_jobs.exp                              FAIL     6.34 s (exit status 1)

--- command ---
SRC_ROOT='/build/ksh/src/ksh-2020.0.0-beta1' SHELL='/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh' TEST_ROOT='/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests' LD_LIBRARY_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libast:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libcmd:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll' LIBSAMPLE_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll/libsample.so' /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh /build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/util/run_test.sh b_jobs.exp
--- Listing only the last 100 lines from a long log. ---
 bg invalid job ID fails
 fg invalid job ID fails
 background job killed when shell exits
-disowned background job not killed when shell exits
-Ctrl-z stops sleep 60 processs
-jobs lists stopped sleep 60 process
-jobs -l lists pid of stopped sleep 60 process
-jobs -p lists process group of stopped sleep 60 process
-bg brings last stopped process to background
-Warning before exit for stopped jobs works
<E> run_test[249]: Stderr for b_jobs.exp should have been empty:
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\nstatus 0\r\n" {
    puts "disowned background job not killed when shell exits"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_jobs.exp" line 97)
<W> run_test[352]: The entire expect's interactive.tmp.log:
<I> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh -l
<D> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh -l: code 0, result 4195
<I> expecting prompt 1
<D> entering expect
<D> expect finished: KSH PROMPT:1: 
<I> send {stty rows 24 cols 80\r}
<I> expecting prompt 2
<D> entering expect
<D> expect finished: stty rows 24 cols 80\r\nstty rows 24 cols 80\r\nKSH PROMPT:2: 
<I> ========== STARTING TEST AT LINE 18 ==========
<I> send {stty -a\r}
<I> expecting prompt 3
<D> entering expect
<D> expect finished: stty -a\r\nspeed 38400 baud; rows 24; columns 80; line = 0;\r\nintr = ^C; quit = ^\\; erase = ^?; kill = ^U; eof = ^D; eol = <undef>;\r\neol2 = <undef>; swtch = <undef>; start = ^Q; stop = ^S; susp = ^Z; rprnt = ^R;\r\nwerase = ^W; lnext = ^V; discard = ^O; min = 1; time = 0;\r\n-parenb -parodd -cmspar cs8 -hupcl -cstopb cread -clocal -crtscts\r\n-ignbrk brkint -ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl ixon -ixoff\r\n-iuclc -ixany imaxbel -iutf8\r\nopost -olcuc -ocrnl onlcr -onocr -onlret -ofill -ofdel nl0 cr0 tab0 bs0 vt0 ff0\r\nisig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt\r\nechoctl echoke -flusho -extproc\r\nKSH PROMPT:3: 
<I> ========== STARTING TEST AT LINE 24 ==========
<I> send {disown 666\r}
<D> entering expect
<D> expect finished: disown 666\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: disown: no such job\r\n
<I> expecting prompt 4
<D> entering expect
<D> expect finished: KSH PROMPT:4: 
<I> ========== STARTING TEST AT LINE 33 ==========
<I> send {bg %9\r}
<D> entering expect
<D> expect finished: bg %9\r\n%9: no such job\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: bg: no such job\r\n
<I> expecting prompt 5
<D> entering expect
<D> expect finished: KSH PROMPT:5: 
<I> ========== STARTING TEST AT LINE 42 ==========
<I> send {fg %9\r}
<D> entering expect
<D> expect finished: fg %9\r\n%9: no such job\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: fg: no such job\r\n
<I> expecting prompt 6
<D> entering expect
<D> expect finished: KSH PROMPT:6: 
<I> ========== STARTING TEST AT LINE 52 ==========
<I> ========== STARTING TEST AT LINE 80 ==========
<I> send {/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh -l\r}
<D> entering expect
<D> expect finished: /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh -l\r\nKSH PROMPT:1:
<I> send {$(type -p sleep) 17 &\r}
<D> entering expect
<D> expect finished:  $(type -p sleep) 17 &\r\n$(type -p sleep) 17 &\r\n[1]\t4204\r\nKSH PROMPT:2:
<I> send {disown %1\r}
<D> entering expect
<D> expect finished:  disown %1\r\nKSH PROMPT:3:
<I> send {ps waux | grep '[s]leep 17'\r}
<D> entering expect
<D> expect finished:  ps waux | grep '[s]leep 17'\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: ps: not found\r\nKSH PROMPT:4:
<I> send {echo status $?\r}
<D> entering expect
<D> expect finished:  echo status $?\r\nstatus 1\r\nKSH PROMPT:5:
<I> send {exit\r}
<I> expecting prompt 7
<D> entering expect
<D> expect finished:  exit\r\nKSH PROMPT:7: 
<I> send {ps waux | grep '[s]leep 17'\r}
<I> expecting prompt 8
<D> entering expect
<D> expect finished: ps waux | grep '[s]leep 17'\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: ps: not found\r\nKSH PROMPT:8: 
<I> send {echo status $?\r}
<D> entering expect
<D> timeout; buffer=echo status $?\r\nstatus 1\r\nKSH PROMPT:9: 
<D> expect returned code 1
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\nstatus 0\r\n" {
    puts "disowned background job not killed when shell exits"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_jobs.exp" line 97)
-------

175/286 b_times.exp                             FAIL    10.65 s (exit status 1)

--- command ---
SRC_ROOT='/build/ksh/src/ksh-2020.0.0-beta1' SHELL='/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh' TEST_ROOT='/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests' LD_LIBRARY_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libast:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libcmd:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll' LIBSAMPLE_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll/libsample.so' /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh /build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/util/run_test.sh b_times.exp
--- Listing only the last 100 lines from a long log. ---
-times produces expected non-zero output for child processes
<E> run_test[249]: Stderr for b_times.exp should have been empty:
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<W> run_test[267]: The last 20 lines of expect's interactive.tmp.log:
<I> ========== STARTING TEST AT LINE 33 ==========
<I> send {{ SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }\r}
<I> expecting prompt 5
<D> entering expect
<D> expect finished: { SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }\r\nKSH PROMPT:5: 
<I> send {times\r}
<D> entering expect
<D> timeout; buffer=times\r\n0m0.847s 0m0.654s\r\n0m0.007s 0m0.000s\r\nKSH PROMPT:6: 
<D> expect returned code 1
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<E> run_test[228]: Stdout for b_times.exp had unexpected differences:
--- /build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp.out	2019-09-06 13:00:13.000000000 -0400
+++ b_times.exp.out	2019-09-06 13:02:19.893530792 -0400
@@ -1,4 +1,2 @@
 times produces expected output
 times sleep 1 produces syntax error
-times produces expected non-zero output for shell
-times produces expected non-zero output for child processes
<E> run_test[249]: Stderr for b_times.exp should have been empty:
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<W> run_test[352]: The entire expect's interactive.tmp.log:
<I> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh
<D> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: code 0, result 5505
<I> expecting prompt 1
<D> entering expect
<D> expect finished: KSH PROMPT:1: 
<I> send {stty rows 24 cols 80\r}
<I> expecting prompt 2
<D> entering expect
<D> expect finished: stty rows 24 cols 80\r\nstty rows 24 cols 80\r\nKSH PROMPT:2: 
<I> ========== STARTING TEST AT LINE 15 ==========
<I> send {times\r}
<D> entering expect
<D> expect finished: times\r\n0m0.004s 0m0.000s\r\n0m0.002s 0m0.003s\r\n
<I> expecting prompt 3
<D> entering expect
<D> expect finished: KSH PROMPT:3: 
<I> ========== STARTING TEST AT LINE 24 ==========
<I> send {times sleep 1\r}
<D> entering expect
<D> expect finished: times sleep 1\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: times: times: unexpected arguments\r\nUsage: times
<I> expecting prompt 4
<D> entering expect
<D> expect finished:  [ options ]\r\nKSH PROMPT:4: 
<I> ========== STARTING TEST AT LINE 33 ==========
<I> send {{ SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }\r}
<I> expecting prompt 5
<D> entering expect
<D> expect finished: { SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }\r\nKSH PROMPT:5: 
<I> send {times\r}
<D> entering expect
<D> timeout; buffer=times\r\n0m0.863s 0m0.636s\r\n0m0.002s 0m0.003s\r\nKSH PROMPT:6: 
<D> expect returned code 1
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
-------
```

**Ksh version:**

2020.0.0beta1

**How reproducible:**

Try updating Arch Linux community/ksh package and building in a clean chroot.

Note: I've already backported the fix for #1390 which was my third testsuite failure.

Trying to build the source tree for git master I still see the same failures.

**Comment by krader1961** [comment]
This statement

```
{ SECONDS=0; while (( SECONDS < 1.5 )); do true; done; }
```

should have consumed at least one second of user mode time. On your system it consumed just  0.863 with the reset of the time spent in system (i.e., kernel) mode. Which is quite odd. Most likely your system isn't using the VDSO mechanism which makes `gettimeofday()` avoid having to make a syscall. But even then the cost of the syscall should be much cheaper, relative to the ksh user space code that is executed, on your system. It certainly is on all the systems I test on. Most of which are virtualized.

Having said that this test is inherently flaky and the way I wrote that specific test makes it more likely to fail even though there is nothing wrong. I'll rework it to make it less flakey and have a more predictable user to system mode time ratio even on systems like yours.

**Comment by krader1961** [comment]
@eli-schwartz Try PR #1399

**Comment by eli-schwartz** [comment]
I tried it but no luck:

```
175/286 b_times.exp                             FAIL     9.80 s (exit status 1)

--- command ---
LD_LIBRARY_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libast:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libcmd:/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll' SHELL='/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh' LIBSAMPLE_PATH='/build/ksh/src/ksh-2020.0.0-beta1/build/src/lib/libdll/libsample.so' TEST_ROOT='/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests' SRC_ROOT='/build/ksh/src/ksh-2020.0.0-beta1' /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh /build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/util/run_test.sh b_times.exp
--- Listing only the last 100 lines from a long log. ---
-times produces expected non-zero output for child processes
<E> run_test[249]: Stderr for b_times.exp should have been empty:
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<W> run_test[267]: The last 20 lines of expect's interactive.tmp.log:
<I> ========== STARTING TEST AT LINE 33 ==========
<I> send {{ SECONDS=0; while (( SECONDS < 1.2 )); do for (( x = 0; x < 10000; x++ )); do true; done; done; }\r}
<I> expecting prompt 5
<D> entering expect
<D> expect finished: { SECONDS=0; while (( SECONDS < 1.2 )); do for (( x = 0; x < 10000; x++ )); do true; done; done; }\r\nKSH PROMPT:5: 
<I> send {times\r}
<D> entering expect
<D> timeout; buffer=times\r\n0m0.909s 0m0.306s\r\n0m0.011s 0m0.000s\r\nKSH PROMPT:6: 
<D> expect returned code 1
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<E> run_test[228]: Stdout for b_times.exp had unexpected differences:
--- /build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp.out	2019-09-07 23:48:08.000000000 -0400
+++ b_times.exp.out	2019-09-07 23:50:19.782158254 -0400
@@ -1,4 +1,2 @@
 times produces expected output
 times sleep 1 produces syntax error
-times produces expected non-zero output for shell
-times produces expected non-zero output for child processes
<E> run_test[249]: Stderr for b_times.exp should have been empty:
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
<W> run_test[352]: The entire expect's interactive.tmp.log:
<I> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh
<D> spawn /build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: code 0, result 5524
<I> expecting prompt 1
<D> entering expect
<D> expect finished: KSH PROMPT:1: 
<I> send {stty rows 24 cols 80\r}
<I> expecting prompt 2
<D> entering expect
<D> expect finished: stty rows 24 cols 80\r\nstty rows 24 cols 80\r\nKSH PROMPT:2: 
<I> ========== STARTING TEST AT LINE 15 ==========
<I> send {times\r}
<D> entering expect
<D> expect finished: times\r\n0m0.002s 0m0.004s\r\n0m0.009s 0m0.001s\r\n
<I> expecting prompt 3
<D> entering expect
<D> expect finished: KSH PROMPT:3: 
<I> ========== STARTING TEST AT LINE 24 ==========
<I> send {times sleep 1\r}
<D> entering expect
<D> expect finished: times sleep 1\r\n/build/ksh/src/ksh-2020.0.0-beta1/build/src/cmd/ksh93/ksh: times: times: unexpected arguments\r\nUsage: times
<I> expecting prompt 4
<D> entering expect
<D> expect finished:  [ options ]\r\nKSH PROMPT:4: 
<I> ========== STARTING TEST AT LINE 33 ==========
<I> send {{ SECONDS=0; while (( SECONDS < 1.2 )); do for (( x = 0; x < 10000; x++ )); do true; done; done; }\r}
<I> expecting prompt 5
<D> entering expect
<D> expect finished: { SECONDS=0; while (( SECONDS < 1.2 )); do for (( x = 0; x < 10000; x++ )); do true; done; done; }\r\nKSH PROMPT:5: 
<I> send {times\r}
<D> entering expect
<D> timeout; buffer=times\r\n0m0.912s 0m0.298s\r\n0m0.009s 0m0.001s\r\nKSH PROMPT:6: 
<D> expect returned code 1
timeout
    while executing
"error $msg"
    (procedure "abort" line 2)
    invoked from within
"abort "timeout""
    invoked from within
"expect -re "\r\n$one_sec_re $zero_sec_re\r\n$zero_sec_re $zero_sec_re\r\n" {
    puts "times produces expected non-zero output for shell"
}"
    (file "/build/ksh/src/ksh-2020.0.0-beta1/src/cmd/ksh93/tests/b_times.exp" line 38)
-------
```

**Comment by eli-schwartz** [comment]
As for why my system is behaving unusually -- I don't think it is. I just checked and the testsuite passes if I run it natively on my system. It is when I run it inside of a systemd-nspawn container that this test fails (actually, both the tests that fail, seem to only do so in the container).

Possibly nspawn containers don't get to use the VDSO mechanism? But you say you tested on mostly virtualized systems; did that include nspawn? I cannot imagine why this would not work inside nspawn anyway...

**Comment by krader1961** [comment]
@eli-schwartz: Your most recent test failure shows that it is very close to passing. The user mode time in the final execution was 0.912 seconds and the test requires it to be >= 1.0 seconds. The ratio of user:system time is much better than your original problem report (by approximately a factor of five). But is still lower than would be expected.

I have no idea what "nspawn" is. My personal test environments include two native macOS systems and several Linux and BSD systems virtualized via VMware. Not to mention the virtualized test environments provided by Travis that are used when pushing changes to Github. There is definitely something unusual about your `nspawn` environment. I am happy to adapt the project so it passes its unit tests in your `nspawn` VM environments. But I have no interest in setting up such an environment and figuring out the solution. So the onus is on you.

**Comment by eli-schwartz** [comment]
[nspawn](https://www.freedesktop.org/software/systemd/man/systemd-nspawn.html) is part of the systemd init, it is not full virtualization but instead "chroot except more powerful". That may or may not be interesting to you, but the links are free. :)

What I did discover, or rather, someone else noticed, is that if systemd itself is built with seccomp support disabled, the test no longer flakes out. As for the seccomp-enabled container:

> it's because of the spinning I think
> it just makes an infinite loop of syscalls, all going through seccomp
> if you look in strace you see it

I guess that answers the "why".

**Comment by krader1961** [comment]
Note that the `b_jobs.exp` test failure is because you don't have the `ps` command in your chroot jail. If you can't install `ps` in your test environments we could check for that command and skip the test if it isn't available. But I'd prefer not to do that since requiring the `ps` command seems like a reasonable dependency.

As for the `b_times.exp` failure I have force pushed a more adaptive approach that should "scale" the busy loop behavior (i.e., how often it calls `gettimeofday()`) based on a quick check of the platform's behavior. This latest commit passes on all my systems and Travis so isn't obviously a step backwards. I ran an instrumented version on a couple of my systems and it certainly looks like it should be an improvement in user:system mode ratio for the tests that are deliberately CPU bound. Whether it is good enough for your test environment is TBD.

**Comment by krader1961** [comment]
Regarding 

> What I did discover, or rather, someone else noticed, is that if systemd itself is built with seccomp support disabled, the test no longer flakes out.

Which makes sense given your description of `nspawn`. It is likely that VDSO is disabled in a "seccomp" enabled jail. Which means `gettimeofday()` is invoked every time `$SECONDS` is expanded.  Minimizing expansion of `$SECONDS` is generally a good idea. And this issue illustrates why.

**Comment by krader1961** [comment]
> I guess if worst comes to worst I can make the testsuite non-fatal (|| warning "flaky tests have failed, check this by hand") and run the testsuite outside of a seccomp-filtered environment?

I'd prefer to either

1) figure out how to make the tests more robust, or
1) skip the problematic test in your environment rather than have you ignore all tests.

Note we already do the latter for other test:platform combinations.

**Comment by eli-schwartz** [comment]
>  that the `b_jobs.exp` test failure is because you don't have the `ps` command in your chroot jail. If

Thanks -- this is not installed by default (in the minimal builder), and I overlooked the part of the log message where it complained about its absence. I should have looked more closely.

I've added it as a "checkdepends", so it will always be available when running the ksh testsuite -- this test now passes. I now have ed, vi, expect, which, netcat, and procps-ng installed as testsuite dependencies, and can easily add more (once I figure out that they are needed).

> As for the `b_times.exp` failure I have force pushed a more adaptive approach that should "scale" the busy loop behavior

I have applied the updated version of #1399 as a patch to my ksh package, and the testsuite now passes. Yay! :)

Thanks -- this looks good to me.
